autoload/lsp/ui/vim/completion.vim

```diff
@@ -0,0 +1,254 @@
+
+  " apply textEdit or insertText(snippet).
+  let l:expand_text = s:get_expand_text(l:completed_item, l:completion_item)
+  if strlen(l:expand_text) > 0
+    call s:clear_inserted_text(
+          \   l:line,
+          \   l:position,
+          \   l:completed_item,
+          \   l:completion_item
+          \ )
+    if exists('g:lsp_snippets_expand_snippet') && len(g:lsp_snippets_expand_snippet) > 0
+      " vim-lsp-snippets expects commit characters removed.
+      call s:expand_text_simply(v:completed_item['word'])
+    elseif exists('g:lsp_snippet_expand') && len(g:lsp_snippet_expand) > 0
+      " other snippet integartion point.
+      call g:lsp_snippet_expand[0]({
+            \   'snippet': l:expand_text
+            \ })
+    else
+      " expand text simply.
+      call s:expand_text_simply(l:expand_text)
+    endif
+  endif
+
+  " apply additionalTextEdits.
+  if has_key(l:completion_item, 'additionalTextEdits')
+    let l:saved_mark = getpos("'a")
+    let l:pos = getpos('.')
+    call setpos("'a", l:pos)
+    call lsp#utils#text_edit#apply_text_edits(
+          \ lsp#utils#get_buffer_uri(bufnr('%')),
+          \ l:completion_item['additionalTextEdits']
+          \ )
+    let l:pos = getpos("'a")
+    call setpos("'a", l:saved_mark)
+    call setpos('.', l:pos)
+  endif
+
+  doautocmd User lsp_complete_done
+  return ''
+endfunction
+
+"
+" Try `completionItem/resolve` if it possible.
+"
+function! s:resolve_completion_item(completion_item, server_name) abort
+  " server_name is not provided.
+  if empty(a:server_name)
+    return a:completion_item
+  endif
+
+  " check server capabilities.
+  let l:capabilities = lsp#get_server_capabilities(a:server_name)
+  if !has_key(l:capabilities, 'completionProvider')
+        \ || !has_key(l:capabilities['completionProvider'], 'resolveProvider')
+        \ || !l:capabilities['completionProvider']['resolveProvider']
+    return a:completion_item
+  endif
+
+  let l:ctx = {}
+  let l:ctx['response'] = {}
+  function! l:ctx['callback'](data) abort
+    let l:self['response'] = a:data.response
+  endfunction
+
+  call lsp#send_request(a:server_name, {
+        \   'method': 'completionItem/resolve',
+        \   'params': a:completion_item,
+        \   'sync': 1,
+        \   'on_notification': function(l:ctx['callback'], [], l:ctx)
+        \ })
+
+  if empty(l:ctx['response'])
+    return a:completion_item
+  endif
+
+  if lsp#client#is_error(l:ctx['response'])
+    return a:completion_item
+  endif
+
+  if empty(l:ctx['response']['result'])
+    return a:completion_item
+  endif
+
+  return l:ctx['response']['result']
+endfunction
+
+"
+" Remove inserted text during completion.
+"
+function! s:clear_inserted_text(line, position, completed_item, completion_item) abort
+  " Remove commit characters.
+  call setline('.', a:line)
+
+  " Create range to remove v:completed_item.
+  let l:range = {
+        \   'start': {
```

Comment:
can we instead create `lsp#utils#position#_vim_to_lsp()` method that does `+1` and `-1` automatically include `to_char` for us? 

```vim
let l:range = {
    \ 'start': lsp#utils#position('%', line1, col1),
    \ 'end': lsp#utils#position('%', line2, col2),
    \ }
```

currently I'm just making sure we have these apis and in future PRs I plan to write proper tests for this so refactoring and adding becomes lot easier for everyone and is consistent even for new users as they don't have to read the entire LSP spec to understand why we need to do `-1` and `+1` sometimes.